### ISSUE TITLE
Data points cannot be written more than 25h10s in the past

### ISSUE THREAD

**Issue by svenmueller** [google/slo-generator#98]
Hi,

The `slo-pipeline` Cloud Function (exports SLO metrics to BQ backend) started to fail suddenly and keeps crashing since then in one of our GCP projects running a slo-pipeline. All logged errors refer to the an invalid interval end time value ("2020-10-20T15:33:00-07:00").

```
google.api_core.exceptions.InvalidArgument: 400 Field timeSeries[0].points[0].interval.end_time had an invalid value of "2020-10-20T15:33:00-07:00": Data points cannot be written more than 25h10s in the past.
```

Full Stacktrace:
```
2020-10-27 23:18:29.552 CET
slo-pipeline2dxhzarq4bqc Traceback (most recent call last): File "/env/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 57, in error_remapped_callable return callable_(*args, **kwargs) File "/env/local/lib/python3.7/site-packages/grpc/_channel.py", line 826, in __call__ return _end_unary_response_blocking(state, call, False, None) File "/env/local/lib/python3.7/site-packages/grpc/_channel.py", line 729, in _end_unary_response_blocking raise _InactiveRpcError(state) grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with: status = StatusCode.INVALID_ARGUMENT details = "Field timeSeries[0].points[0].interval.end_time had an invalid value of "2020-10-20T15:18:00-07:00": Data points cannot be written more than 25h10s in the past." debug_error_string = "{"created":"@1603837109.545536881","description":"Error received from peer ipv4:74.125.133.95:443","file":"src/core/lib/surface/call.cc","file_line":1061,"grpc_message":"Field timeSeries[0].points[0].interval.end_time had an invalid value of "2020-10-20T15:18:00-07:00": Data points cannot be written more than 25h10s in the past.","grpc_status":3}" > The above exception was the direct cause of the following exception: Traceback (most recent call last): File "/env/local/lib/python3.7/site-packages/google/cloud/functions/worker_v2.py", line 449, in run_background_function _function_handler.invoke_user_function(event_object) File "/env/local/lib/python3.7/site-packages/google/cloud/functions/worker_v2.py", line 268, in invoke_user_function return call_user_function(request_or_event) File "/env/local/lib/python3.7/site-packages/google/cloud/functions/worker_v2.py", line 265, in call_user_function event_context.Context(**request_or_event.context)) File "/user_code/main.py", line 25, in main compute.export(data, exporters) File "/env/local/lib/python3.7/site-packages/slo_generator/compute.py", line 98, in export ret = exporter.export(data, **config) File "/env/local/lib/python3.7/site-packages/slo_generator/exporters/stackdriver.py", line 50, in export self.create_timeseries(data, **config) File "/env/local/lib/python3.7/site-packages/slo_generator/exporters/stackdriver.py", line 93, in create_timeseries result = self.client.create_time_series(project, [series]) File "/env/local/lib/python3.7/site-packages/google/cloud/monitoring_v3/gapic/metric_service_client.py", line 1060, in create_time_series request, retry=retry, timeout=timeout, metadata=metadata File "/env/local/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 145, in __call__ return wrapped_func(*args, **kwargs) File "/env/local/lib/python3.7/site-packages/google/api_core/retry.py", line 286, in retry_wrapped_func on_error=on_error, File "/env/local/lib/python3.7/site-packages/google/api_core/retry.py", line 184, in retry_target return target() File "/env/local/lib/python3.7/site-packages/google/api_core/timeout.py", line 214, in func_with_timeout return func(*args, **kwargs) File "/env/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable six.raise_from(exceptions.from_grpc_error(exc), exc) File "<string>", line 3, in raise_from google.api_core.exceptions.InvalidArgument: 400 Field timeSeries[0].points[0].interval.end_time had an invalid value of "2020-10-20T15:18:00-07:00": Data points cannot be written more than 25h10s in the past. 
```

Any idea why we see this and how we can fix it?

**Comment by ocervell** [comment]
Hi Sven,

I haven't encountered this one yet, but `2020-10-20T15:33:00-07:00` doesn't seem like an UTC timestamp, which Stackdriver probably expects: by any chance did you change anything related to time zones in Cloud Scheduler / Cloud Functions code ?

**Comment by igorpeshansky** [comment]
The timestamp comes directly from the event data payload, so it might just be a delayed pubsub message or something (whatever the trigger for your function was). If you run with debug logging enabled, you should see the timestamp printed from various places in the code.

**Comment by ocervell** [comment]
@svenmueller I think this is related to running Cloud Scheduler in US and the GCF code being in EU ... Currently we fetch the timestamp for the scheduler. 
Is this a transient issue or a permanent one ? 
Did you happen see those errors after changing the app engine region ?

**Comment by igorpeshansky** [comment]
The formatting of the timestamp in the error message uses PST just because that's how the Monitoring API formats it, so that's probably a red herring.
If this is a transient issue, you should be able to easily reproduce it by explicitly sending a Pub/Sub message to the same channel with a stale timestamp.

**Comment by svenmueller** [comment]
> @svenmueller I think this is related to running Cloud Scheduler in US and the GCF code being in EU ... Currently we fetch the timestamp for the scheduler.

We had to go with Cloud Scheduler in US region since it was already set as default for this GCP project (seems like the default app engine region can't be changed).

> Is this a transient issue or a permanent one ?
Since then, the slo-pipeline CF keeps crashing (logging the mentioned log event above) permanently.

> Did you happen see those errors after changing the app engine region ?
As far as i know,  the default app engine region can't be changed (there was no change from our side).

**Comment by ocervell** [comment]
@svenmueller can you please try upgrading to `terraform-google-slo` version `v0.3.2` and check if you have the same behaviour ?

**Comment by svenmueller** [comment]
After migrating to the latest stable version, the slo-pipeline CF stopped crashing (not sure if it will appear again after some time as it did the first appeared without know reason). I will close this issue and re-open it as soon as it appears again.

**Comment by ocervell** [comment]
Re-opening since seen in 2 customer environments

**Comment by ocervell** [comment]
Terraform module `2.0.0` will handle this by notably replacing Cloud Functions with Cloud Run (much more stable), and will be released at the same time as `slo-generator` v2.0 (mid-August ETA).

**Comment by ocervell** [comment]
Moved to https://github.com/terraform-google-modules/terraform-google-slo/issues/90
